File: WEEK 1/DAY 5 - Linked Lists Advanced/SOLUTIONS/03_Hard_Problem_Solution/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        import heapq
        
        min_heap = []
        
        # Add the head of each list to the heap
        for l in lists:
            if l:
                heapq.heappush(min_heap, (l.val, l))
                
        dummy = ListNode(0)
        current = dummy
        
        while min_heap:
            val, node = heapq.heappop(min_heap)
            current.next = node
            current = current.next
            if node.next:
                heapq.heappush(min_heap, (node.next.val, node.next))
                
        return dummy.next

File: WEEK 1/DAY 5 - Linked Lists Advanced/SOLUTIONS/03_Hard_Problem_Solution/Solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function(lists) {
    let heap = new MinPriorityQueue({ priority: x => x.val });
    
    for (let list of lists) {
        if (list) heap.enqueue(list);
    }
    
    let dummy = new ListNode(0);
    let current = dummy;
    
    while (!heap.isEmpty()) {
        let node = heap.dequeue().element;
        current.next = node;
        current = current.next;
        if (node.next) heap.enqueue(node.next);
    }
    
    return dummy.next;
};
